File: packages/sdk-typescript/src/integrations/openai.ts
/**
 * wrapOpenAI — instrument an OpenAI client so every model call emits a Causal
 * `llm` span carrying the model, the prompt/completion io, and the tokens and
 * cost read straight off `response.usage`.
 *
 * The client is accepted structurally (duck-typed), so `@causal/sdk` never
 * takes a dependency on `openai` — the same wrapper works for Azure OpenAI,
 * OpenRouter, Together and anything else that speaks the OpenAI shape.
 *
 * Usage:
 *   import OpenAI from "openai";
 *   import { CausalTracer } from "@causal/sdk";
 *   import { wrapOpenAI } from "@causal/sdk/integrations";
 *
 *   const tracer = new CausalTracer({ service: "support-agent" });
 *
 *   await tracer.trace("support_agent.run", async (t, root) => {
 *     const openai = wrapOpenAI(new OpenAI(), t, { parent: root });
 *
 *     // one `llm` span, nested under the run's root span
 *     const res = await openai.chat.completions.create({
 *       model: "gpt-4o-mini",
 *       messages: [{ role: "user", content: "Where is my order?" }],
 *     });
 *     return res.choices[0]?.message.content ?? "";
 *   });
 *
 * Streaming works too. Pass `stream_options: { include_usage: true }` and the
 * span picks up tokens and cost from the final chunk:
 *
 *   const stream = await openai.chat.completions.create({
 *     model: "gpt-4o-mini",
 *     messages,
 *     stream: true,
 *     stream_options: { include_usage: true },
 *   });
 *   for await (const chunk of stream) process.stdout.write(chunk.choices[0]?.delta?.content ?? "");
 *   // the span ends when the loop ends — including on `break` or a throw
 */

import {
  attrs,
  buildIo,
  errMessage,
  estimateCost,
  isPromiseLike,
  num,
  openSpan,
  pick,
  pickPath,
  proxyPath,
  str,
  stringify,
  traceStream,
  type SpanTarget,
  type WrapOptions,
} from "./internal.js";

/* eslint-disable @typescript-eslint/no-explicit-any -- proxy plumbing is inherently untyped */
type AnyFn = (...args: any[]) => any;

/* ── prompt / completion rendering ──────────────────────────────────── */

/** Render one message's `content`, which may be a string or a parts array. */
function renderContent(content: unknown): string {
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content
      .map((part) => {
        const type = pick(part, "type");
        if (type === "text" || type === "input_text" || type === "output_text") {
          return String(pick(part, "text") ?? "");
        }
        if (type === "refusal") return `[refusal] ${String(pick(part, "refusal") ?? "")}`;
        return `[${String(type ?? "part")}]`;
      })
      .join("");
  }
  if (content === undefined || content === null) return "";
  return stringify(content);
}

/** Flatten a chat `messages` array into a readable transcript. */
function renderMessages(messages: unknown): string {
  if (!Array.isArray(messages)) return stringify(messages);
  return messages
    .map((message) => {
      const role = String(pick(message, "role") ?? "user");
      const calls = pick(message, "tool_calls");
      const body = renderContent(pick(message, "content"));
      return calls ? `${role}: ${body}\n[tool_calls] ${stringify(calls)}` : `${role}: ${body}`;
    })
    .join("\n");
}

/** Render whatever the Responses API was handed as `input`. */
function renderResponsesInput(params: unknown): string {
  const instructions = str(pick(params, "instructions"));
  const input = pick(params, "input");
  const rendered =
    typeof input === "string"
      ? input
      : Array.isArray(input)
        ? input
            .map((item) => {
              const role = pick(item, "role");
              const body = renderContent(pick(item, "content"));
              return role ? `${String(role)}: ${body}` : body || stringify(item);
            })
            .join("\n")
        : stringify(input);
  return instructions ? `[instructions] ${instructions}\n${rendered}` : rendered;
}

/** Pull the assistant text out of a Responses API response object. */
function renderResponsesOutput(response: unknown): string {
  const convenience = pick(response, "output_text");
  if (typeof convenience === "string") return convenience;
  const output = pick(response, "output");
  if (!Array.isArray(output)) return stringify(output);
  return output
    .map((item) => {
      const type = pick(item, "type");
      if (type === "message") return renderContent(pick(item, "content"));
      if (type === "function_call") {
        return `[function_call ${String(pick(item, "name") ?? "")}] ${String(pick(item, "arguments") ?? "")}`;
      }
      return `[${String(type ?? "item")}]`;
    })
    .filter((line) => line.length > 0)
    .join("\n");
}

/* ── usage ─────────────────────────────────────────────────────────── */

interface Usage {
  tokensIn: number;
  tokensOut: number;
  cachedTokens?: number;
  reasoningTokens?: number;
}

/** Chat Completions reports `prompt_tokens` / `completion_tokens`. */
function readChatUsage(usage: unknown): Usage | undefined {
  const tokensIn = num(pick(usage, "prompt_tokens"));
  const tokensOut = num(pick(usage, "completion_tokens"));
  if (tokensIn === undefined && tokensOut === undefined) return undefined;
  return {
    tokensIn: tokensIn ?? 0,
    tokensOut: tokensOut ?? 0,
    cachedTokens: num(pickPath(usage, "prompt_tokens_details", "cached_tokens")),
    reasoningTokens: num(pickPath(usage, "completion_tokens_details", "reasoning_tokens")),
  };
}

/** The Responses API reports `input_tokens` / `output_tokens`. */
function readResponsesUsage(usage: unknown): Usage | undefined {
  const tokensIn = num(pick(usage, "input_tokens"));
  const tokensOut = num(pick(usage, "output_tokens"));
  if (tokensIn === undefined && tokensOut === undefined) return undefined;
  return {
    tokensIn: tokensIn ?? 0,
    tokensOut: tokensOut ?? 0,
    cachedTokens: num(pickPath(usage, "input_tokens_details", "cached_tokens")),
    reasoningTokens: num(pickPath(usage, "output_tokens_details", "reasoning_tokens")),
  };
}

/* ── the wrappers ──────────────────────────────────────────────────── */

/** Mutable state a streamed call accumulates before the span can be closed. */
interface Accumulator {
  text: string;
  usage?: unknown;
  finishReason?: string;
  model?: string;
  responseId?: string;
}

function instrument(
  orig: AnyFn,
  target: SpanTarget,
  options: WrapOptions,
  spec: {
    defaultName: string;
    renderInput: (params: unknown) => string;
    renderOutput: (value: unknown) => string;
    readUsage: (usage: unknown) => Usage | undefined;
    usageOf: (value: unknown) => unknown;
    onChunk: (acc: Accumulator, chunk: unknown) => void;
  }
): AnyFn {
  return function create(...args: unknown[]): unknown {
    const params = args[0];
    const requestedModel = str(pick(params, "model"));
    const streaming = pick(params, "stream") === true;
    const span = openSpan(target, options.spanName ?? spec.defaultName, "llm", options.parent);
    const input = spec.renderInput(params);

    const close = (
      output: string,
      usage: Usage | undefined,
      extra: Record<string, unknown>,
      error?: unknown
    ): void => {
      const model = str(extra["model"]) ?? requestedModel;
      const cost = usage ? estimateCost(model, usage.tokensIn, usage.tokensOut, options.prices) : undefined;
      const end: Parameters<typeof span.end>[0] = {
        status: error ? "error" : "ok",
        attributes: attrs({
          provider: "openai",
          model,
          stream: streaming ? "true" : "false",
          temperature: num(pick(params, "temperature")),
          max_tokens: num(pick(params, "max_tokens")) ?? num(pick(params, "max_output_tokens")),
          "cache.read_tokens": usage?.cachedTokens,
          "reasoning.tokens": usage?.reasoningTokens,
          ...extra,
        }),
      };
      const io = buildIo(options, input, output);
      if (io) end.io = io;
      if (usage) {
        end.tokensIn = usage.tokensIn;
        end.tokensOut = usage.tokensOut;
        if (cost !== undefined) end.cost = cost;
      }
      if (error) end.error = errMessage(error);
      span.end(end);
      void span.finish();
    };

    const onValue = (value: unknown): void => {
      const usage = spec.readUsage(spec.usageOf(value));
      close(spec.renderOutput(value), usage, {
        model: str(pick(value, "model")),
        "response.id": str(pick(value, "id")),
        finish_reason: str(pickPath(value, "choices", "0", "finish_reason")),
      });
    };

    const onStream = (stream: object): object => {
      const acc: Accumulator = { text: "" };
      return traceStream(
        stream,
        (chunk) => spec.onChunk(acc, chunk),
        (err) =>
          close(acc.text, spec.readUsage(acc.usage), {
            model: acc.model,
            "response.id": acc.responseId,
            finish_reason: acc.finishReason,
          }, err)
      );
    };

    let result: unknown;
    try {
      result = orig(...args);
    } catch (err) {
      close("", undefined, {}, err);
      throw err;
    }

    if (isPromiseLike(result)) {
      if (!streaming) {
        // Observe without replacing the returned value: the OpenAI SDK hands
        // back an APIPromise, and callers rely on `.withResponse()` / `.asResponse()`.
        void Promise.resolve(result).then(onValue, (err: unknown) => close("", undefined, {}, err));
        return result;
      }
      // Streaming: we must hand back an instrumented stream, so this call site
      // resolves to a plain Promise. Streams have no APIPromise extras to lose.
      return Promise.resolve(result).then(
        (value: unknown) => (value && typeof value === "object" ? onStream(value) : value),
        (err: unknown) => {
          close("", undefined, {}, err);
          throw err;
        }
      );
    }

    if (streaming && result && typeof result === "object") return onStream(result);
    onValue(result);
    return result;
  };
}

/** Accumulate a `chat.completions` stream chunk. */
function onChatChunk(acc: Accumulator, chunk: unknown): void {
  const model = str(pick(chunk, "model"));
  if (model) acc.model = model;
  const id = str(pick(chunk, "id"));
  if (id) acc.responseId = id;
  const usage = pick(chunk, "usage");
  if (usage) acc.usage = usage;

  const choices = pick(chunk, "choices");
  if (!Array.isArray(choices)) return;
  for (const choice of choices) {
    const delta = pick(choice, "delta");
    const text = pick(delta, "content");
    if (typeof text === "string") acc.text += text;
    const calls = pick(delta, "tool_calls");
    if (Array.isArray(calls)) {
      for (const call of calls) {
        const name = str(pickPath(call, "function", "name"));
        const chunkArgs = pickPath(call, "function", "arguments");
        if (name) acc.text += `\n[tool_call ${name}] `;
        if (typeof chunkArgs === "string") acc.text += chunkArgs;
      }
    }
    const finish = str(pick(choice, "finish_reason"));
    if (finish) acc.finishReason = finish;
  }
}

/** Accumulate a Responses API stream event. */
function onResponsesEvent(acc: Accumulator, event: unknown): void {
  const type = str(pick(event, "type")) ?? "";
  if (type === "response.output_text.delta") {
    const delta = pick(event, "delta");
    if (typeof delta === "string") acc.text += delta;
    return;
  }
  if (type === "response.function_call_arguments.delta") {
    const delta = pick(event, "delta");
    if (typeof delta === "string") acc.text += delta;
    return;
  }
  const response = pick(event, "response");
  if (!response) return;
  const model = str(pick(response, "model"));
  if (model) acc.model = model;
  const id = str(pick(response, "id"));
  if (id) acc.responseId = id;
  const usage = pick(response, "usage");
  if (usage) acc.usage = usage;
  const status = str(pick(response, "status"));
  if (status) acc.finishReason = status;
  if (!acc.text) {
    const text = renderResponsesOutput(response);
    if (text) acc.text = text;
  }
}

/**
 * Wrap an OpenAI-shaped client so `chat.completions.create` and
 * `responses.create` emit `llm` spans.
 *
 * The original client is never mutated — a Proxy is returned, so an
 * un-instrumented reference to the same client keeps working and two tracers
 * can wrap the same client independently.
 *
 * @param client  An `OpenAI` (or Azure/compatible) instance.
 * @param target  A `CausalTracer`, a live `CausalTrace`, or a parent `CausalSpan`.
 *                Prefer a trace or span: a bare tracer opens and flushes a
 *                one-span trace per call.
 * @param options Span naming, io capture limits and price-table overrides.
 */
export function wrapOpenAI<T extends object>(client: T, target: SpanTarget, options: WrapOptions = {}): T {
  if (!client || typeof client !== "object") return client;

  const withChat = proxyPath(client, ["chat", "completions", "create"], (orig) =>
    instrument(orig, target, options, {
      defaultName: "openai.chat.completions.create",
      renderInput: (params) => renderMessages(pick(params, "messages")),
      renderOutput: (value) => {
        const message = pickPath(value, "choices", "0", "message");
        const content = renderContent(pick(message, "content"));
        const calls = pick(message, "tool_calls");
        if (calls) return content ? `${content}\n[tool_calls] ${stringify(calls)}` : stringify(calls);
        return content;
      },
      readUsage: readChatUsage,
      usageOf: (value) => pick(value, "usage"),
      onChunk: onChatChunk,
    })
  );

  return proxyPath(withChat, ["responses", "create"], (orig) =>
    instrument(orig, target, options, {
      defaultName: "openai.responses.create",
      renderInput: renderResponsesInput,
      renderOutput: renderResponsesOutput,
      readUsage: readResponsesUsage,
      usageOf: (value) => pick(value, "usage"),
      onChunk: onResponsesEvent,
    })
  );
}

File: packages/sdk-python/causal_sdk/integrations/openai.py
"""
wrap_openai — instrument an OpenAI client so every model call emits a Causal
`llm` span carrying the model, the prompt/completion io, and the tokens and
cost read straight off `response.usage`.

The client is accepted structurally (duck-typed), so `causal-sdk` never takes a
dependency on `openai` — the same wrapper works for `OpenAI`, `AsyncOpenAI`,
`AzureOpenAI` and anything else that speaks the OpenAI shape.

Usage:

    from openai import OpenAI
    from causal_sdk import CausalTracer
    from causal_sdk.integrations.openai import wrap_openai

    tracer = CausalTracer(service="support-agent")

    async def main() -> None:
        t = tracer.start_trace()
        root = t.span("support_agent.run", "agent")

        client = wrap_openai(OpenAI(), t, parent=root)

        # one `llm` span, nested under the run's root span
        res = client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[{"role": "user", "content": "Where is my order?"}],
        )

        root.end(status="ok")
        await t.flush()

Streaming works too. Pass ``stream_options={"include_usage": True}`` and the
span picks up tokens and cost from the final chunk; the span closes when the
loop ends, including on ``break`` or an exception.
"""

from __future__ import annotations

import functools
import inspect
from collections.abc import Mapping
from typing import Any, Callable

from ._tracing import (
    ModelPrice,
    SpanHandle,
    SpanTarget,
    TracedProxy,
    TracedStream,
    as_int,
    as_str,
    attrs,
    build_io,
    err_message,
    estimate_cost,
    num,
    open_span,
    pick,
    pick_path,
    stringify,
)

__all__ = ["wrap_openai"]


# ── prompt / completion rendering ─────────────────────────────────────


def _render_content(content: Any) -> str:
    """Render one message's `content`, which may be a string or a parts list."""
    if isinstance(content, str):
        return content
    if content is None:
        return ""
    if not isinstance(content, (list, tuple)):
        return stringify(content)
    parts: list[str] = []
    for part in content:
        part_type = pick(part, "type")
        if part_type in ("text", "input_text", "output_text"):
            parts.append(str(pick(part, "text") or ""))
        elif part_type == "refusal":
            parts.append(f"[refusal] {pick(part, 'refusal') or ''}")
        else:
            parts.append(f"[{part_type or 'part'}]")
    return "".join(parts)


def _render_messages(messages: Any) -> str:
    """Flatten a chat `messages` list into a readable transcript."""
    if not isinstance(messages, (list, tuple)):
        return stringify(messages)
    lines: list[str] = []
    for message in messages:
        role = pick(message, "role") or "user"
        body = _render_content(pick(message, "content"))
        calls = pick(message, "tool_calls")
        lines.append(
            f"{role}: {body}\n[tool_calls] {stringify(calls)}" if calls else f"{role}: {body}"
        )
    return "\n".join(lines)


def _render_responses_input(params: Mapping[str, Any]) -> str:
    """Render whatever the Responses API was handed as `input`."""
    instructions = as_str(params.get("instructions"))
    value = params.get("input")
    if isinstance(value, str):
        rendered = value
    elif isinstance(value, (list, tuple)):
        lines = []
        for item in value:
            role = pick(item, "role")
            body = _render_content(pick(item, "content"))
            lines.append(f"{role}: {body}" if role else (body or stringify(item)))
        rendered = "\n".join(lines)
    else:
        rendered = stringify(value)
    return f"[instructions] {instructions}\n{rendered}" if instructions else rendered


def _render_responses_output(response: Any) -> str:
    """Pull the assistant text out of a Responses API response object."""
    convenience = pick(response, "output_text")
    if isinstance(convenience, str):
        return convenience
    output = pick(response, "output")
    if not isinstance(output, (list, tuple)):
        return stringify(output)
    lines: list[str] = []
    for item in output:
        item_type = pick(item, "type")
        if item_type == "message":
            lines.append(_render_content(pick(item, "content")))
        elif item_type == "function_call":
            name = pick(item, "name") or ""
            lines.append(f"[function_call {name}] {pick(item, 'arguments') or ''}")
        else:
            lines.append(f"[{item_type or 'item'}]")
    return "\n".join(line for line in lines if line)


def _render_chat_output(response: Any) -> str:
    choices = pick(response, "choices")
    first = choices[0] if isinstance(choices, (list, tuple)) and choices else None
    message = pick(first, "message")
    content = _render_content(pick(message, "content"))
    calls = pick(message, "tool_calls")
    if calls:
        return f"{content}\n[tool_calls] {stringify(calls)}" if content else stringify(calls)
    return content


# ── usage ─────────────────────────────────────────────────────────────


class _Usage:
    __slots__ = ("tokens_in", "tokens_out", "cached_tokens", "reasoning_tokens")

    def __init__(
        self,
        tokens_in: int,
        tokens_out: int,
        cached_tokens: int | None = None,
        reasoning_tokens: int | None = None,
    ) -> None:
        self.tokens_in = tokens_in
        self.tokens_out = tokens_out
        self.cached_tokens = cached_tokens
        self.reasoning_tokens = reasoning_tokens


def _read_chat_usage(usage: Any) -> _Usage | None:
    """Chat Completions reports `prompt_tokens` / `completion_tokens`."""
    tokens_in = num(pick(usage, "prompt_tokens"))
    tokens_out = num(pick(usage, "completion_tokens"))
    if tokens_in is None and tokens_out is None:
        return None
    return _Usage(
        as_int(tokens_in),
        as_int(tokens_out),
        cached_tokens=_optional_int(pick_path(usage, "prompt_tokens_details", "cached_tokens")),
        reasoning_tokens=_optional_int(
            pick_path(usage, "completion_tokens_details", "reasoning_tokens")
        ),
    )


def _read_responses_usage(usage: Any) -> _Usage | None:
    """The Responses API reports `input_tokens` / `output_tokens`."""
    tokens_in = num(pick(usage, "input_tokens"))
    tokens_out = num(pick(usage, "output_tokens"))
    if tokens_in is None and tokens_out is None:
        return None
    return _Usage(
        as_int(tokens_in),
        as_int(tokens_out),
        cached_tokens=_optional_int(pick_path(usage, "input_tokens_details", "cached_tokens")),
        reasoning_tokens=_optional_int(
            pick_path(usage, "output_tokens_details", "reasoning_tokens")
        ),
    )


def _optional_int(value: Any) -> int | None:
    number = num(value)
    return int(number) if number is not None else None


# ── stream accumulation ───────────────────────────────────────────────


class _Accumulator:
    __slots__ = ("text", "usage", "finish_reason", "model", "response_id")

    def __init__(self) -> None:
        self.text = ""
        self.usage: Any = None
        self.finish_reason: str | None = None
        self.model: str | None = None
        self.response_id: str | None = None


def _on_chat_chunk(acc: _Accumulator, chunk: Any) -> None:
    acc.model = as_str(pick(chunk, "model")) or acc.model
    acc.response_id = as_str(pick(chunk, "id")) or acc.response_id
    usage = pick(chunk, "usage")
    if usage is not None:
        acc.usage = usage

    choices = pick(chunk, "choices")
    if not isinstance(choices, (list, tuple)):
        return
    for choice in choices:
        delta = pick(choice, "delta")
        text = pick(delta, "content")
        if isinstance(text, str):
            acc.text += text
        calls = pick(delta, "tool_calls")
        if isinstance(calls, (list, tuple)):
            for call in calls:
                name = as_str(pick_path(call, "function", "name"))
                if name:
                    acc.text += f"\n[tool_call {name}] "
                arguments = pick_path(call, "function", "arguments")
                if isinstance(arguments, str):
                    acc.text += arguments
        finish = as_str(pick(choice, "finish_reason"))
        if finish:
            acc.finish_reason = finish


def _on_responses_event(acc: _Accumulator, event: Any) -> None:
    event_type = as_str(pick(event, "type")) or ""
    if event_type in ("response.output_text.delta", "response.function_call_arguments.delta"):
        delta = pick(event, "delta")
        if isinstance(delta, str):
            acc.text += delta
        return
    response = pick(event, "response")
    if response is None:
        return
    acc.model = as_str(pick(response, "model")) or acc.model
    acc.response_id = as_str(pick(response, "id")) or acc.response_id
    usage = pick(response, "usage")
    if usage is not None:
        acc.usage = usage
    acc.finish_reason = as_str(pick(response, "status")) or acc.finish_reason
    if not acc.text:
        acc.text = _render_responses_output(response)


# ── the wrapper ───────────────────────────────────────────────────────


class _Spec:
    """Per-endpoint rendering/usage strategy."""

    __slots__ = ("default_name", "render_input", "render_output", "read_usage", "on_chunk")

    def __init__(
        self,
        default_name: str,
        render_input: Callable[[Mapping[str, Any]], str],
        render_output: Callable[[Any], str],
        read_usage: Callable[[Any], _Usage | None],
        on_chunk: Callable[[_Accumulator, Any], None],
    ) -> None:
        self.default_name = default_name
        self.render_input = render_input
        self.render_output = render_output
        self.read_usage = read_usage
        self.on_chunk = on_chunk


def _merge_params(args: tuple[Any, ...], kwargs: dict[str, Any]) -> dict[str, Any]:
    params: dict[str, Any] = {}
    if args and isinstance(args[0], Mapping):
        params.update(args[0])
    params.update(kwargs)
    return params


def _instrument(
    original: Callable[..., Any],
    target: SpanTarget,
    spec: _Spec,
    *,
    parent: Any,
    span_name: str | None,
    capture_io: bool,
    max_io_chars: int,
    prices: Mapping[str, ModelPrice] | None,
) -> Callable[..., Any]:
    is_async = inspect.iscoroutinefunction(original)

    def start(params: Mapping[str, Any]) -> tuple[SpanHandle, str]:
        handle = open_span(target, span_name or spec.default_name, "llm", parent=parent)
        return handle, spec.render_input(params)

    def close(
        handle: SpanHandle,
        params: Mapping[str, Any],
        input_text: str,
        output_text: str,
        usage: _Usage | None,
        extra: Mapping[str, Any],
        error: BaseException | None = None,
    ) -> None:
        model = as_str(extra.get("model")) or as_str(params.get("model"))
        payload: dict[str, Any] = {
            "status": "error" if error is not None else "ok",
            "attributes": attrs(
                {
                    "provider": "openai",
                    "model": model,
                    "stream": "true" if params.get("stream") else "false",
                    "temperature": params.get("temperature"),
                    "max_tokens": params.get("max_tokens") or params.get("max_output_tokens"),
                    "cache.read_tokens": usage.cached_tokens if usage else None,
                    "reasoning.tokens": usage.reasoning_tokens if usage else None,
                    **extra,
                }
            ),
        }
        io = build_io(input_text, output_text, capture_io=capture_io, max_io_chars=max_io_chars)
        if io:
            payload["io"] = io
        if usage is not None:
            payload["tokens_in"] = usage.tokens_in
            payload["tokens_out"] = usage.tokens_out
            payload["cost"] = estimate_cost(model, usage.tokens_in, usage.tokens_out, prices)
        if error is not None:
            payload["error"] = err_message(error)
        handle.end(**payload)
        handle.finish()

    def close_value(
        handle: SpanHandle, params: Mapping[str, Any], input_text: str, value: Any
    ) -> None:
        choices = pick(value, "choices")
        first = choices[0] if isinstance(choices, (list, tuple)) and choices else None
        close(
            handle,
            params,
            input_text,
            spec.render_output(value),
            spec.read_usage(pick(value, "usage")),
            {
                "model": as_str(pick(value, "model")),
                "response.id": as_str(pick(value, "id")),
                "finish_reason": as_str(pick(first, "finish_reason")),
            },
        )

    def wrap_stream(
        handle: SpanHandle, params: Mapping[str, Any], input_text: str, stream: Any
    ) -> TracedStream:
        acc = _Accumulator()
        return TracedStream(
            stream,
            lambda chunk: spec.on_chunk(acc, chunk),
            lambda error: close(
                handle,
                params,
                input_text,
                acc.text,
                spec.read_usage(acc.usage),
                {
                    "model": acc.model,
                    "response.id": acc.response_id,
                    "finish_reason": acc.finish_reason,
                },
                error,
            ),
        )

    if is_async:

        @functools.wraps(original)
        async def acreate(*args: Any, **kwargs: Any) -> Any:
            params = _merge_params(args, kwargs)
            handle, input_text = start(params)
            try:
                result = await original(*args, **kwargs)
            except BaseException as exc:
                close(handle, params, input_text, "", None, {}, exc)
                raise
            if params.get("stream"):
                return wrap_stream(handle, params, input_text, result)
            close_value(handle, params, input_text, result)
            return result

        return acreate

    @functools.wraps(original)
    def create(*args: Any, **kwargs: Any) -> Any:
        params = _merge_params(args, kwargs)
        handle, input_text = start(params)
        try:
            result = original(*args, **kwargs)
        except BaseException as exc:
            close(handle, params, input_text, "", None, {}, exc)
            raise
        if params.get("stream"):
            return wrap_stream(handle, params, input_text, result)
        close_value(handle, params, input_text, result)
        return result

    return create


_CHAT_SPEC = _Spec(
    "openai.chat.completions.create",
    lambda params: _render_messages(params.get("messages")),
    _render_chat_output,
    _read_chat_usage,
    _on_chat_chunk,
)

_RESPONSES_SPEC = _Spec(
    "openai.responses.create",
    _render_responses_input,
    _render_responses_output,
    _read_responses_usage,
    _on_responses_event,
)


def wrap_openai(
    client: Any,
    target: SpanTarget,
    *,
    parent: Any = None,
    span_name: str | None = None,
    capture_io: bool = True,
    max_io_chars: int = 4000,
    prices: Mapping[str, ModelPrice] | None = None,
) -> Any:
    """Wrap an OpenAI-shaped client so its model calls emit `llm` spans.

    Both `OpenAI` and `AsyncOpenAI` are supported — the wrapper inspects the
    underlying method and produces a matching sync or async callable.

    The original client is never mutated: a forwarding proxy is returned, so an
    un-instrumented reference to the same client keeps working and two tracers
    can wrap the same client independently.

    Args:
        client: An `OpenAI` / `AsyncOpenAI` / `AzureOpenAI` instance.
        target: A `CausalTracer`, a live trace, or a parent span. Prefer a trace
            or span; a bare tracer opens and flushes a one-span trace per call.
        parent: Optional parent span to nest the emitted spans under.
        span_name: Override the span name (default: the provider method).
        capture_io: Record prompt/completion text in `span.io`. Default `True`.
        max_io_chars: Truncate each side of `span.io` to this many characters.
        prices: Extra per-1M-token prices, merged over the built-in table.

    Returns:
        A proxy that behaves exactly like `client`.
    """
    if client is None:
        return client

    options = {
        "parent": parent,
        "span_name": span_name,
        "capture_io": capture_io,
        "max_io_chars": max_io_chars,
        "prices": prices,
    }

    overrides: dict[str, Any] = {}

    completions = getattr(getattr(client, "chat", None), "completions", None)
    if completions is not None and callable(getattr(completions, "create", None)):
        traced_completions = TracedProxy(
            completions,
            {"create": _instrument(completions.create, target, _CHAT_SPEC, **options)},
        )
        overrides["chat"] = TracedProxy(client.chat, {"completions": traced_completions})

    responses = getattr(client, "responses", None)
    if responses is not None and callable(getattr(responses, "create", None)):
        overrides["responses"] = TracedProxy(
            responses,
            {"create": _instrument(responses.create, target, _RESPONSES_SPEC, **options)},
        )

    if not overrides:
        return client
    return TracedProxy(client, overrides)
